src/shims/x86/sse2.rs

```diff
@@ -0,0 +1,1026 @@
+                    let left = this.read_scalar(&this.project_index(&left, i)?)?.to_u8()?;
+                    let right = this.read_scalar(&this.project_index(&right, i)?)?.to_u8()?;
+                    let dest = this.project_index(&dest, i)?;
+
+                    // Values are expanded from u8 to u16, so adds cannot overflow.
+                    let res = u16::from(left).wrapping_add(u16::from(right)).wrapping_add(1) >> 1;
+                    this.write_scalar(Scalar::from_u8(res.try_into().unwrap()), &dest)?;
+                }
+            }
+            // Used to implement the _mm_avg_epu16 function.
+            // Averages packed unsigned 16-bit integers in `left` and `right`.
+            "pavg.w" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, left_len);
+                assert_eq!(dest_len, right_len);
+
+                for i in 0..dest_len {
+                    let left = this.read_scalar(&this.project_index(&left, i)?)?.to_u16()?;
+                    let right = this.read_scalar(&this.project_index(&right, i)?)?.to_u16()?;
+                    let dest = this.project_index(&dest, i)?;
+
+                    // Values are expanded from u16 to u32, so adds cannot overflow.
+                    let res = u32::from(left).wrapping_add(u32::from(right)).wrapping_add(1) >> 1;
+                    this.write_scalar(Scalar::from_u16(res.try_into().unwrap()), &dest)?;
+                }
+            }
+            // Used to implement the _mm_mulhi_epi16 function.
+            "pmulh.w" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(dest_len, left_len);
+                assert_eq!(dest_len, right_len);
+
+                for i in 0..dest_len {
+                    let left = this.read_scalar(&this.project_index(&left, i)?)?.to_i16()?;
+                    let right = this.read_scalar(&this.project_index(&right, i)?)?.to_i16()?;
+                    let dest = this.project_index(&dest, i)?;
+
+                    // Values are expanded from i16 to i32, so multiplication cannot overflow.
+                    let res = i32::from(left).wrapping_mul(i32::from(right)) >> 16;
+                    this.write_scalar(Scalar::from_int(res, Size::from_bits(16)), &dest)?;
+                }
+            }
+            // Used to implement the _mm_mulhi_epu16 function.
+            "pmulhu.w" => {
```

Comment:
I wonder if you could de-duplicate some code here by sharing this with `pmulh.w`, and using operations on `ImmTy` (as produced by `read_immediate`) instead of host integer arithmetic? We already have all arithmetic operations implemented in `overflowing_binary_op` after all.

This applies throughout the entire file. There seems to be a lot of use of host integers here and that's generally surprising. (And same for floats, `overflowing_binary_op` will do the right thing on immediates of float type as well.)

Reply:
Thanks for trying this out! Looks like the shared version is overall longer than the 2 separate versions... so I guess it's not worth it.

Are there other sets of intrinsics that are completely the same except for the types involved?

Reply:
Like, for instance pavg also exists twice. With some helper methods for widening casts, could this be useful?

"No I think that would introduce more overhead than it helps" is a reasonable answer here, I'm just asking you since it's hard to say for myself without actually trying to write that code. :)

Reply:
It might be worth trying it out, almost half of the loop body is used by the widening casts.

But I would leave it for the future PR you mentioned below (the one about making `bin_op_f32`/`bin_op_f64` generic).